File: java/src/main/java/com/dsa/examples/searching/LinearSearch.java
// Searching — Linear Search

package com.dsa.examples.searching;

/**
 * Linear search over an array.
 * Time: O(n) per search; Space: O(1).
 */
public final class LinearSearch {

    private LinearSearch() {
    }

    /**
     * Returns the index of {@code key} in {@code arr}, or -1 if absent.
     * Time: O(n); Space: O(1).
     */
    public static int indexOf(int[] arr, int key) {
        if (arr == null) {
            return -1;
        }
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == key) {
                return i;
            }
        }
        return -1;
    }
}

File: python/dsa/searching/linear_search.py
# Searching — Linear Search

"""Linear search over an array. Time: O(n) per search; Space: O(1)."""


def index_of(arr: list[int] | None, key: int) -> int:
    """Returns the index of key in arr, or -1 if absent.

    Time: O(n); Space: O(1).
    """
    if arr is None:
        return -1
    for i, value in enumerate(arr):
        if value == key:
            return i
    return -1
